### PR TITLE
🌱 Allow passing in provider contract and clusterctl binary in the clusterctl upgrade spec

### PR DESCRIPTION
**Pull request by randomvariable**
Signed-off-by: Naadir Jeewa <jeewan@vmware.com>

**What this PR does / why we need it**:
I would like to run both v1alpha4 -> v1beta1 and v1alpha3 -> v1beta1 tests in the same job, so expanding the clusterctl upgrade spec to allow different provider contracts to be tested.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/assign @sbueringer 
/area testing

### CODE REVIEW

File: test/e2e/clusterctl_upgrade.go [hunk 1]
```diff
@@ -143,8 +152,7 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		// Get a ClusterProxy so we can interact with the workload cluster
 		managementClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(ctx, cluster.Namespace, cluster.Name)
 
-		// Download the v1alpha3 clusterctl version to be used for setting up the management cluster to be upgraded
-		clusterctlBinaryURL := input.E2EConfig.GetVariable(initWithBinaryVariableName)
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply test/e2e/clusterctl_upgrade.go:147]: Moved. Also switched to strings.NewReplacer. The functionalist in me doesn't like modifying string vars.

File: test/e2e/clusterctl_upgrade.go [hunk 2]
```diff
@@ -90,6 +93,12 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
 		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
 		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		if input.InitWithBinary == "" {
+			Expect(input.E2EConfig.Variables).To(HaveKey(initWithBinaryVariableName), "Invalid argument. %s variable must be defined when calling %s spec", initWithBinaryVariableName, specName)
+			clusterctlBinaryURL = input.E2EConfig.GetVariable(initWithBinaryVariableName)
+		} else {
+			clusterctlBinaryURL = input.InitWithBinary
+		}
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply test/e2e/clusterctl_upgrade.go:103]: Yes, missed that one. Sorted.

**Comment by sbueringer** [comment]
/lgtm

Pending successful tests obviously. But the errors look like they are caused by a Github problem.

/test pull-cluster-api-e2e-full-main
(to also run the clusterctl test)

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**